bin/git/git.js

```diff
@@ -0,0 +1,243 @@
+      throw new Error(stripIndents`
+        Unable to execute git command '${command}'.
+
+        Chromatic only works in git projects.
+        Contact us at support@hichroma.com if you need to use Chromatic outside of one.
+      `);
+    }
+
+    if (message.match('does not have any commits yet')) {
+      throw new Error(
+        `Unable to execute git command '${command}'.
+
+Chromatic requires that you have created a commit before it can be run.
+`
+      );
+    }
+
+    throw error;
+  }
+}
+
+export const FETCH_N_INITIAL_BUILD_COMMITS = 20;
+
+const TesterFirstCommittedAtQuery = gql`
+  query TesterFirstCommittedAtQuery($branch: String!) {
+    app {
+      firstBuild(sortByCommittedAt: true) {
+        committedAt
+      }
+      lastBuild(branch: $branch, sortByCommittedAt: true) {
+        commit
+        committedAt
+      }
+    }
+  }
+`;
+
+const TesterHasBuildsWithCommitsQuery = gql`
+  query TesterHasBuildsWithCommitsQuery($commits: [String!]!) {
+    app {
+      hasBuildsWithCommits(commits: $commits)
+    }
+  }
+`;
+
+// NOTE: At some point we should check that the commit has been pushed to the
+// remote and the branch matches with origin/REF, but for now we are naive about
+// adhoc builds.
+
+// We could cache this, but it's probably pretty quick
+export async function getCommit() {
+  const [commit, committedAtSeconds, committerEmail, committerName] = (await execGitCommand(
+    `git log -n 1 --format="%H,%ct,%ce,%cn"`
+  )).split(',');
+
+  return { commit, committedAt: committedAtSeconds * 1000, committerEmail, committerName };
+}
+
+export async function getBranch() {
+  return execGitCommand(`git rev-parse --abbrev-ref HEAD`);
+}
+
+// Check if a commit exists in the repository
+async function commitExists(commit) {
+  try {
+    await execGitCommand(`git cat-file -e "${commit}^{commit}"`);
+    return true;
+  } catch (error) {
+    return false;
+  }
+}
+
+function commitsForCLI(commits) {
+  return commits.map(c => c.trim()).join(' ');
+}
+
+// git rev-list in a basic form gives us a list of commits reaching back to
+// `firstCommittedAtSeconds` (i.e. when the first build of this app happened)
+// in reverse chronological order.
+//
+// A simplified version of what we are doing here is just finding the first
+// commit in that list that has a build. We only want to send `limit` to
+// the server in this pass (although we may already know some commits that do
+// or do not have builds from earlier passes). So we just pick the first `limit`
+// commits from the command, filtering out `commitsWith[out]Builds`.
+//
+// However, it's not quite that simple -- because of branching. However,
+// passing commits after `--not` in to `git rev-list` *occludes* all the ancestors
+// of those commits. This is exactly what we need once we find one or more commits
+// that do have builds: a list of the ancestors of HEAD that are not accestors of
+// `commitsWithBuilds`.
+//
+async function nextCommits(
```

Comment:
Maybe we should pull this baseline calculation stuff into a separate branch

Reply:
that sounds like something we could do for V4?

Reply:
Sure, whatever works